k8_cortx_cloud/rolling-upgrade-cortx-cloud.sh

```diff
@@ -0,0 +1,254 @@
+            ready_count="${ready_counts[0]}"
+            total_count="${ready_counts[1]}"
+            if [[ -n ${pod_name} ]]; then
+                if [[ ${pod_status} != "Running" ]]; then
+                    printf "  %s %s -> status is %s\n" "${RED}✘${CLEAR}" "${pod_name}" "${pod_status}"
+                    pods_ready=false
+                elif (( ready_count != total_count )); then
+                    printf "  %s %s -> only %s/%s pods are ready\n" "${RED}✘${CLEAR}" "${pod_name}" "${ready_count}" "${total_count}"
+                    pods_ready=false
+                else
+                    printf "  %s %s\n" "${GREEN}✓${CLEAR}" "${pod_name}"
+                fi
+            fi
+        done <<< "${cortx_pods}"
+    fi
+    printf "\n"
+
+    if [[ ${pods_ready} == false ]]; then
+        printf "Pre-upgrade pod readiness check failed. Ensure all pods are in a healthy state, or manually shutdown the cluster, and try again.\n"
+        exit 1
+    fi
+
+}
+
+function upgrade_control_pod() {
+    # Upgrade Control Pod
+    cortxcontrol_image=$(parse_solution 'solution.images.cortxcontrol' | cut -f2 -d'>')
+
+    printf "Pausing Control pod..\n";
+    kubectl rollout pause deployment cortx-control;
+
+    printf "Updating Control pod configuration..\n";
+    kubectl set env deployment/cortx-control UPGRADE_TYPE="Rolling" --containers="*";
+    kubectl set image deployment cortx-control "*=${cortxcontrol_image}";
+
+    printf "Resuming Control pod..\n";
+    kubectl rollout resume deployment cortx-control;
+    sleep $TIMEDELAY;
+
+    printf "Control Pod Upgraded\n";
+}
+
+function upgrade_ha_pod() {
+    # Upgrad HA Pod
+    cortxha_image=$(parse_solution 'solution.images.cortxha' | cut -f2 -d'>')
+
+    printf "Pausing HA pod..\n";
+    kubectl rollout pause deployment cortx-ha;
+
+    printf "updating HA pod configuration..\n";
+    kubectl set env deployment/cortx-ha UPGRADE_TYPE="Rolling" --containers="*";
+    kubectl set image deployment cortx-ha "*=${cortxha_image}";
+
+    printf "Resuming HA pod..\n";
+    kubectl rollout resume deployment cortx-ha;
+
+    sleep $TIMEDELAY;
+}
+
+function upgrade_data_pod() {
+    # Upgrad Data Pods
+    cortxdata_image=$(parse_solution 'solution.images.cortxdata' | cut -f2 -d'>')
+
+    while IFS= read -r line; do
+        IFS=" " read -r -a deployments <<< "$line"
+
+        printf "Pausing "${deployments[0]}" pod..\n";
+        kubectl rollout pause deployment "${deployments[0]}";
+
+        printf "Updating "${deployments[0]}" pod configuration..\n";
+        kubectl set image deployment "${deployments[0]}"  "*=${cortxdata_image}";
+        kubectl set env deployment/"${deployments[0]}" UPGRADE_TYPE="Rolling" --containers="*";
+
+        printf "Resuming "${deployments[0]}" pod..\n";
+        kubectl rollout resume deployment "${deployments[0]}";
+
+        sleep $TIMEDELAY;
+    done <<< "$(kubectl get deployments |grep 'cortx-data-')"
+}
+
+function upgrade_server_pod() {
```

Comment:
Lot of common code between all upgrade_xxx_pod(). Better to create a common function and call it from all upgrade functions, e.g. 

```
function upgrade_pod() { 
    pod_name=$1
    ...
}
function upgrade_server_pod() {
    ....
    upgrade_pod "server"
    ....
}
```

Reply:
done